python/lsst/daf/butler/core/connectionString.py

```diff
@@ -0,0 +1,97 @@
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+__all__ = ("DB_AUTH_ENVVAR", "DB_AUTH_PATH", "ConnectionStringFactory")
+
+from sqlalchemy.engine import url
+from lsst.daf.butler.core.dbAuth import DbAuth
+from lsst.daf.butler.core.registry import RegistryConfig
+
+DB_AUTH_ENVVAR = "LSST_DB_AUTH"
+"""Default name of the environmental variable that will be used to locate DB
+credentials configuration file. """
+
+DB_AUTH_PATH = "~/.lsst/db-auth.yaml"
+"""Default path at which it is expected that DB credentials are found."""
+
+
+class ConnectionStringFactory:
+    """Factory for `sqlalchemy.engine.url.URL` instances.
+
+    The factory constructs a connection string URL object by parsing the
+    connection string, the 'db' key in the registry configuration.
+    Username, password, host, port or database can be specified as keys in the
+    config explicitly. If username or password are missing a matching DB is
+    found in the credentials file pointed to by `DB_AUTH_ENVVAR` or
+    `DB_AUTH_PATH` values.
+    """
+
+    keys = ('username', 'password', 'host', 'port', 'database')
+
+    @classmethod
+    def fromConfig(cls, registryConfig):
+        """Parses the 'db' key in the config, and if they exist username,
+        password, host, port and database keys, and returns an connection
+        string object.
+
+        If no  username and password are found in the connection string, or in
+        the config, they are retrieved from a file at `DB_AUTH_PATH` or
+        `DB_AUTH_ENVVAR`. Sqlite dialect does not require a password.
+
+        Parameters
+        ----------
+        config : `ButlerConfig`, `RegistryConfig`, `Config` or `str`
+            Registry configuration
+
+        Returns
+        -------
+        connectionString : `sqlalchemy.engine.url.URL`
+            URL object representing the connection string.
+
+
+        Raises
+        ------
+        DBAuthError
+            If the credentials file has incorrect permissions, doesn't exist at
+            the given location or is formatted incorrectly.
+        """
+        regConf = RegistryConfig(registryConfig)
+        conStr = url.make_url(regConf['db'])
+
+        for key in cls.keys:
+            if getattr(conStr, key) is None:
+                setattr(conStr, key, regConf.get(key))
+
+        # when the databsase is a file, host will be None and the path
+        # becomes the database, but we are not treating 'localhost' explicitly
+        host = conStr.host
+        if conStr.host is None:
+            host = str(conStr.database)
```

Comment:
I still dislike this.  If there is no host, we shouldn't set one, especially not to the database name.  I'd just `return conStr` here.

Reply:
This is a relic from before when I used to let the `db` string be just the host or a path and the `cls` key to set the dialect. If the `db` was a path then it would become the database name, but if it was an url, the host. Since I was treating the host as the target location previously I had to make that switch. Since we kicked out `cls` and I discovered URL (thanks to Andy) I believe it's not needed now at all so I removed it. 

This affects the username and password setting issue I asked you about. I don't want to return the connection string here because there are single file databases that support username and passwords (LiteDB, RaptorDB, sqlite has an extension that supports usernames and passwords). I don't know/think that they would be ever used but it's been made clear to me that I should try to be as general as possible (but also that it's ok not to do it for sqlite). 

So I left it as is, but I take it, from the early `return`, that you're not fine with not using DbAuth when username+password are given in db connection string or as explicit keys in the config file for single file dbs?